KAFKA-13351: Add possibility to write kafka headers in Kafka Console Producer

florin-akermann opened the pull request:
Just like with the ConsoleConsumer, if headers and/or key enabled then the order is "headers key value".
The separator for the header and headerKey are configurable. The defaults are the same as in the ConsoleConsumer.

In short,

Default parsing pattern 

when:
parse.headers=true and parse.key=true:
    "h1:v1,h2...\tkey\tvalue"

when:
parse.headers=false and parse.key=true:
    "key\tvalue"

when:
parse.headers=true and parse.key=false:
    "h1:v1,h2...\tvalue" 

Testing strategy:
The testing is constrained to testing LineMessageReader as this is the only affected class.
Different combinations of parse.headers=true|false and parse.key=true|false are tested.
Different string patterns are tested. Including input that does not fit the expected Pattern.
LineMessageReaderTest.testLineReader suggests that the change is backward compatible.

This contribution is my original work and I license the work to the project under the project's open source license.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

florin-akermann committed:
KAFKA-13351: Add possibility to write kafka headers in Kafka Console Producer
- extended readLine capability of ConsoleProducer.LineMessageReader
- added unit test for ConsoleProducer.LineMessageReader

florin-akermann commented:
@cmccabe

florin-akermann commented:
Hi,

Regarding the failing tests: I think the kafka.admin.LeaderElectionCommandTest might contain flaky tests. They pass locally. See image. Could you trigger a rerun?  I am asking as it says to do so in [here](https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes).

![image](https://user-images.githubusercontent.com/17885538/139603373-77bfc1e8-85bb-4dd1-a1f7-cbbbaaf775b1.png)

dajac commented:
@florin-akermann Thanks for the PR! I think that we need a small KIP for this change cause it introduces new properties. It is similar to https://cwiki.apache.org/confluence/display/KAFKA/KIP-431%3A+Support+of+printing+additional+ConsumerRecord+fields+in+DefaultMessageFormatter which introduces new properties but to the console-consumer.

florin-akermann commented:
@dajac Thanks for the feedback. Ok, I signed up on that confluence page. Unfortunatley I do not have the privileges to create a KIP. Could you grant me the privilege to do so?

dajac commented:
@florin-akermann What's your username? I will do it tomorrow.

florin-akermann commented:
@dajac great, thanks. My username is 'florin.akermann'.

dajac commented:
@florin-akermann done!

florin-akermann commented:
@dajac Thanks. I have aligned the property naming with KIP-431 in the latest commit. Here is the KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-798%3A+Add+possibility+to+write+kafka+headers+in+Kafka+Console+Producer

What would be the next step in the process?

dajac commented:
@florin-akermann The process is described here: https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals (Process chapter). Basically, you have to kick off a DISCUSS thread in the mailing list to discuss your proposal with the community. Then the KIP must be voted on and accepted.

florin-akermann commented:
@dajac Thanks for the guidance so far. The KIP seems to be approved - thanks for the vote :).
What else is needed to bring this across the finish line?

dajac commented:
@florin-akermann Awesome! I will try to make a pass on the PR in the next few days.

florin-akermann commented:
@dajac i dare to nag :) or shall I nag on dev@kafka.apache.org instead?

dajac commented:
@florin-akermann Sorry for the delay... Your PR is still in my review queue. I've been quite busy with the AK 3.1 release (blockers, validation, and co).

florin-akermann commented:
@dajac Ok thanks, good to know it is queued.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -274,28 +296,74 @@ object ConsoleProducer {
+        (line, parseKey, parseHeader) match {
+          case (null, _, _) => null
+          case (line, true, true) =>
+            val Array(headers, key, value) = line.split("[" + headersDelimiter + keySeparator + "]")
```
The downsides of using `split` with a regex are that it has to compile the regular expression for every single line, allocate an array, and deconstruct the array. That seems to be a bit wasteful here given that we only have to split two times. Did you consider using two `indexOf` in conjunction with `substring` to extract the header part, the key part, and finally the value part?

florin-akermann replied:
Good point. I pushed an alternative based on your suggestion.
For the mapHeaders method I compile two regex patterns on init as the number of the key value pairs in the header varies.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -208,9 +213,22 @@ object ConsoleProducer {
-      "\tparse.key=true|false\n" +
-      "\tkey.separator=<key.separator>\n" +
-      "\tignore.error=true|false")
+      "\tparse.key=false\n" +
+      "\tparse.headers=false\n" +
+      "\tignore.error=false\n" +
+      "Default parsing pattern when:\n" +
+      "\tparse.headers=true & parse.key=true:\n" +
+      "\t \"h1:v1,h2...\\tkey\\tvalue\"\n" +
+      "\tparse.headers=false & parse.key=true:\n" +
+      "\t \"key\\tvalue\"\n" +
+      "\tparse.headers=true & parse.key=false:\n" +
+      "\t \"h1:v1,h2...\\tvalue\"\n" +
+      "Customize pattern via (defaults shown)\n" +
+      "\tkey.separator=\\t\n" +
+      "\theaders.delimiter=\\t\n" +
+      "\theaders.separator=,\n" +
+      "\theaders.key.separator=:"
+      )
```
This is quite hard to read. You could use a multiline string in Scala with `"""   """`. You can take a look at how we did it in the `ConsoleConsumer.`

Also, it would be great if we could have all the parameters on the top followed by the examples.

florin-akermann replied:
Adjusted

dajac commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,202 @@
+
+  @Test
+  def testLineReader(): Unit = {
+    val lineReader = new LineMessageReader();
```
I really like all the tests! However, they are quite repetitive. One way to reduce the boilerplate code would be to define a helper method which runs the test case. The method would take the input line, the properties, and the expected results.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -274,28 +296,74 @@ object ConsoleProducer {
+    }
+
+    def mapHeaders(headers: String): lang.Iterable[Header] = {
+      asList(
+        headers.split(headersSeparator)
+          .map(_.split(headerKeySeparator))
+          .map(keyValue => new RecordHeader(keyValue(0), keyValue(1).getBytes(StandardCharsets.UTF_8)))
+          : _*
+      )
```
What does it do if the header is malformed?

florin-akermann replied:
Uff, thanks for the catch.
Now all exceptions get caught and the expected pattern is displayed in the exceptions' message.
`testMalformedHeader` 'documents' the behavior.

dajac reviewed (commented):
@florin-akermann Thanks for the PR! I left a few general comments below.

florin-akermann committed:
KAFKA-13351:
- rm some boilerplate from test
- fix and test case for malformed header

florin-akermann committed:
KAFKA-13351:
- use multiline string feature

florin-akermann committed:
KAFKA-13351:
- avoiding string.split() replaced with combination of indexOf and substring
- compiled regex pattern for header strings where number of elements varies

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,38 +282,113 @@ object ConsoleProducer {
+      if (ignoreError) {
+        new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
+      } else {
+        throw new KafkaException("Could not parse line " + lineNumber + ", most likely line does not match pattern: " + illustratePattern())
```
We can use string interpolation here: `KafkaException(s"Could not parse line $lineNumber ...` rather than concatenation

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -206,11 +210,25 @@ object ConsoleProducer {
+        | headers.key.separator=:
+        |Default parsing pattern when:
+        | parse.headers=true & parse.key=true:
+        |  "h1:v1,h2...\tkey\tvalue"
```
Should we have `h1:v1,h2:v2...\tkey\tvalue` ?
Same below

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -206,11 +210,25 @@ object ConsoleProducer {
+        | headers.separator=,
+        | headers.key.separator=:
+        |Default parsing pattern when:
+        | parse.headers=true & parse.key=true:
```
Should we use `and` rather than `&` ?

florin-akermann replied:
Now i remember the reason why I chose &
![snap](https://user-images.githubusercontent.com/17885538/145125783-ae0e418a-cf2f-4b83-8ba8-d83df3fb9b9e.png)

The lines break in an unfortunate way if 'and' is used. What do you think?

mimaison replied:
Maybe it's time to use more than 80 columns! But that's a discussion for another KIP.

As all these `parse.*` default to `false`, what about only listing the configs set to `true` in the example?
For example:
`parse.headers=false & parse.key=true` -> `parse.key=true`

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,38 +282,113 @@ object ConsoleProducer {
+    }
+
+    private def parseHeaderKeyValue(line: String): ProducerRecord[Array[Byte],Array[Byte]] = {
+      val headerIndex = line.indexOf(headersDelimiter)
```
It seems that the whole parsing is less robust than it was before? What would happen if the delimiter is not found? I think that `indexOf` returns `-1` in this case. I suppose that an exception is thrown and caught in `readMessage`. Am I getting this right?

I wonder if it would be better to be explicit instead. This would also give use the opportunity to provide more precise errors to the user. What do you think?

florin-akermann replied:
Yes you got it right.

I agree, a more precise error message will be useful. Moreover, if the user runs it with ignore.error=true it will be expensive to throw and catch frequently.

So if the user sets ignore.error=true we will no longer throw and catch. Instead the last resort is to send the whole line as a value.

If ignore.errors=false then we throw a single time with the most precise error message possible.

Now I also realise that this principle is applied in the current version on trunk.

In short, I'll adjust. Thanks again for the input.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,38 +282,113 @@ object ConsoleProducer {
+      try{
+        (line, parseKey, parseHeader) match {
+          case (null, _, _) => null
+          case (line, true, true) => parseHeaderKeyValue(line)
+          case (line, true, false) => parseKeyValue(line)
+          case (line, false, true) => parseHeaderValue(line)
+          case (line, false, false) => parseValue(line)
```
Instead of doing this, did you consider have one method? We know that we always have the headers, then the key, and finally the value so we could just skip some part based on the selected options.

florin-akermann replied:
Yes, I think this will fit well with the above discussed adjustment. Change incoming...

dajac reviewed (commented):
@florin-akermann Thanks for the update. I left a few more comments/questions.

mimaison commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,193 @@
+      "headerKey0.0:headerValue0.0&headerKey0.1:headerValue0.1!key0#value0\n" +
+        "headerKey1.0:headerValue1.0!key1#value1"
+
+    val headers: lang.Iterable[Header] = asList(
```
Should this be named `headers0`?

mimaison commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,193 @@
+
+    val input =
+      "headerKey0.0:headerValue0.0&headerKey0.1:headerValue0.1!key0#value0\n" +
+        "headerKey1.0:headerValue1.0!key1#value1"
```
Can we align it with the line above?
Same in a few tests below

mimaison commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,193 @@
+    )
+
+    val record0 = new ProducerRecord("topic", null, null, "key0", "value0", headers)
+
```
We can remove this empty line. Same for the line after `runTest()` below

mimaison reviewed (commented):
Thanks for the PR. It look good overall, I left a few minor suggestions

florin-akermann committed:
KAFKA-13351:
- rm match error catch in favor of more robust code / checks

florin-akermann commented:
Hi again,

@dajac 
@mimaison 

I made the adjustments based on your inputs.

florin-akermann committed:
KAFKA-13351:
- refactoring

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -29,8 +25,16 @@ import kafka.utils.{CommandDefaultOptions, CommandLineUtils, Exit, ToolsUtils}
 import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
 import org.apache.kafka.common.KafkaException
+import org.apache.kafka.common.header.Header
+import org.apache.kafka.common.header.internals.RecordHeader
```
Ideally we should avoid using this internal class. Can we instead use https://kafka.apache.org/30/javadoc/org/apache/kafka/common/header/Headers.html?

florin-akermann replied:
@mimaison  Hmmm, thanks, isn't it the case that RecordHeader is the only implementation of Headers?

Or in other words what implementation should I return on line 353?
https://github.com/apache/kafka/pull/11456/files#diff-3b7393aaaffd466b3bfc53c8d95109a206442160b878bbdaa0b0332257afaebaR353

mimaison replied:
The alternative is to create a `ProducerRecord` and then do `record.headers.add(String, byte[])`

mimaison reviewed (commented):
Thanks for the updates, it looks good. I left one more suggestion.

dajac commented:
@florin-akermann Sorry for the delay. I was not able to do any reviews last week. I will do a pass over it today or tomorrow.

mimaison commented:
A point that was raised in the discussion on KIP-810 is whether we should prevent setting `headers.delimiter`, `headers.separator` and `headers.key.separator` to colliding values.

Commit:
KAFKA-13351: check invariants: header demarcations may not be equal

florin-akermann commented:
> A point that was raised in the discussion on KIP-810 is whether we should prevent setting `headers.delimiter`, `headers.separator` and `headers.key.separator` to colliding values.

@mimaison Thanks for the flag. I commited a check for it.

@dajac, @mimaison  What are your thoughts on this? I am wondering whether this behavior can be added without stating it explicitly in the KIP?  I would say yes.  It would be hard for the user to provide valid input (apart from some corner cases) when 2 out of the three configs in question are equal.

mimaison commented:
@florin-akermann I don't think you need to update the KIP. Thanks for pushing an update

Commit:
KAFKA-13351: rm use of header.internals

Commit:
KAFKA-13351: extend default parsing pattern example

florin-akermann committed:
KAFKA-13351: rm use of header.internals in test class

florin-akermann committed:
KAFKA-13351: adjust indentation of multiline string

florin-akermann committed:
KAFKA-13351: minor - adjust grouping of lines

florin-akermann committed:
KAFKA-13351: 'and' instead of '&'

florin-akermann commented:
Hi again @dajac @mimaison,

Thanks again for all the great input.
I believe the current version has taken into account all your comments.

I am in no hurry, just want to make sure this PR is still in your review queue.

mimaison commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,210 @@
+  }
+
+  @Test
+  def testMissingHeaderSeparator(): Unit = {
```
Shouldn't it be `testMissingKeySeparator()` instead?

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+    }
   }
 }
+
```
These newlines can be deleted

mimaison commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+            if (value != null) value.getBytes(StandardCharsets.UTF_8) else null,
+          )
+
+          if (headers != null){
```
nit: Let's add a space before the opening bracket

mimaison reviewed (commented):
Thanks for the update. I'll looks good!
I left a couple of very minor suggestions, I'm happy to approve once these are addressed.

dajac commented:
I’ll make a pass tomorrow. Sorry for the delay. I could not do it sooner due to the holidays.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+        case line =>
+
+          val headers = parse(parseHeader, line, headersDelimiter, "headers delimiter")
+          val key = parse(parseKey, line.substring(offset(headers)), keySeparator, "key separator")
```
nit: Instead of using `substring` here and below, we could pass the start index to `indexOf`.

florin-akermann replied:
Thanks, I do not follow. Would you prefer to pass the index of the starting char as the second parameter for parse(..., index, ... )? In this case parse() would need to have acces to 'line'? Which in turn would mean an additional parameter to parse()? Moreover, wouldn't this then mean the start index needs to passed as well.

An adavantage of this might be a more detailed error messages as the full malformed line can be added to the exception instead of just the segment of the line and line number.

I guess an other option could be to completley inline parse() at the cost of a 'somewhat' redundant match expression.

Anyhow, before we go any further on this: As it is 'just' a 'nit' I dare asking whether it is ok to leave as is?

dajac replied:
Yeah, this is exactly what I thought about. You're right. We would need to pass both the `line` and the start index to `parse`. For the headers' call, it would be `0` all the time. For the key's call, it would be `offset(headers)`.

This is basically a minor optimization, avoiding to unnecessary allocate a `String` with `line.substring(offset(headers))` as we could work on `line` all the time. I don't feel strong about this so I leave it up to you.

florin-akermann replied:
fair enough, I adjusted it

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+      line match {
+        case null => null
+        case line =>
+
```
nit: We could remove this empty line.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+
+    private def splitHeaders(headers: String): Array[(String, Array[Byte])] = {
+      headerSeparatorPattern.split(headers)
+        .map(pair =>
```
nit: Could we move `map` to the previous line and use curly braces instead of parenthesis? That would be more align with our code style.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+          })
+    }
+
+    private def offset(headers: String) = {
```
nit: Should we rename `headers` as we use this method for both `headers` and `key`?

dajac commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,210 @@
+
+  //  The equality method of ProducerRecord compares memory references for the header iterator, this is why this custom equality check is used.
+  private def assertEquality[K, V](expected: ProducerRecord[K, V], actual: ProducerRecord[Array[Byte], Array[Byte]]): Unit = {
+    assertEquals(expected.key(), if (actual.key() == null) null else new String(actual.key()))
+    assertEquals(expected.value(), if (actual.value() == null) null else new String(actual.value()))
+    assertEquals(expected.headers().toArray.toList, actual.headers().toArray.toList)
```
nit: We could omit the parenthesis for all the getters here (e.g. `key`, `value`, `headers`). We usually don't put them for getters in our code base. There might be a few other cases in other method. I haven't checked them all.

dajac commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,210 @@
+  @Test
+  def testIgnoreErrorInInput(): Unit = {
+    val input =
+      "headerKey0.0:headerValue0.0,headerKey0.1:headerValue0.1[MISSING-DELIMITER]key0\tvalue0\n" +
```
Should we also add cases for missing delimiter in the headers part? For instance, a missing `:` or a missing `,` or even both at the same time.

dajac reviewed (commented):
@florin-akermann Thanks for the update and again excuse me for the delay. Overall, it looks pretty good. I left a few small comments for consideration.

florin-akermann committed:
KAFKA-13351:
 - testMissingHeaderSeparator -> testMissingKeySeparator
 - testMalformedHeader -> testMissingHeaderKeySeparator
 - additional test input for testIgnoreErrorInInput

florin-akermann committed:
KAFKA-13351: rm '()' from side effect free arity-0 methods

florin-akermann committed:
KAFKA-13351: align with code style

florin-akermann committed:
KAFKA-13351: order imports alphabetically, rm empty lines

florin-akermann committed:
KAFKA-13351: align with code-style: convert mapping to block expression

mimaison reviewed (approved):
LGTM! Thanks @florin-akermann for the KIP and PR, this is a nice improvement to the console producer.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +277,89 @@ object ConsoleProducer {
+
+          if (headers != null) {
+            splitHeaders(headers)
+              .foreach(header => record.headers().add(header._1, header._2))
```
nit: `()` could be omitted here as well.

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +277,89 @@ object ConsoleProducer {
+    }
   }
-}
+}
```
nit: Could we add a new line at the end of the file like it was before?

dajac commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  //  The equality method of ProducerRecord compares memory references for the header iterator, this is why this custom equality check is used.
+  private def assertEquality[K, V](expected: ProducerRecord[K, V], actual: ProducerRecord[Array[Byte], Array[Byte]]): Unit = {
```
nit: Would it make sense to call this one `assertRecordEquals`?

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +277,89 @@ object ConsoleProducer {
+        headerKeySeparator = props.getProperty("headers.key.separator")
       if (props.containsKey("ignore.error"))
         ignoreError = props.getProperty("ignore.error").trim.equalsIgnoreCase("true")
+      if (headersDelimiter.equals(headersSeparator))
```
nit: In Scala, we can use `==` to compare strings. It is a bit more readable than using `equals`.

dajac commented on core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala:
```diff
@@ -0,0 +1,226 @@
+ * limitations under the License.
+ */
+
+package unit.kafka.tools
```
nit: I think that we usually don't put the `unit` prefix for the package. Could you check other classes in the same package?

dajac commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -264,39 +277,89 @@ object ConsoleProducer {
-              if (ignoreError) new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
-              else throw new KafkaException(s"No key found on line $lineNumber: $line")
-            case n =>
-              val value = (if (n + keySeparator.size > line.size) "" else line.substring(n + keySeparator.size)).getBytes(StandardCharsets.UTF_8)
```
* It seems that we supported separator of arbitrary size before whereas now we always assume 1.
* I am curious about this check: `if (n + keySeparator.size > line.size) ""`. We don't have it anymore with your changes. Do we still need it?

florin-akermann replied:
point 1: Good catch! I adjusted it and added a unit test for it.
point 2: Maybe it is just too late in the evening but I cannot see how this if statement could ever evaluate to true.

dajac reviewed (commented):
@florin-akermann Thanks for the update. We are almost there. I left a few more comments.

florin-akermann committed:
KAFKA-13351: addressing nits

florin-akermann committed:
KAFKA-13351: fix case for demarcations of length > 1

florin-akermann committed:
KAFKA-13351: rm use of substring

florin-akermann committed:
KAFKA-13351: adjust package name

florin-akermann commented:
> LGTM! Thanks @florin-akermann for the KIP and PR, this is a nice improvement to the console producer.

@mimaison Thanks a lot for a the PR review. The whole process was a nice experience.

dajac reviewed (approved):
LGTM! @florin-akermann Thanks for your contribution!

florin-akermann commented:
@dajac Thanks a lot for the PR review and the guidance from A-Z. Just like with Mickael, the whole process was a nice experience.
